File: packages/sdk/python/examples/ai_inference.py
import asyncio
import os
from solders.keypair import Keypair
from x402_upl import SolanaX402Client, X402Config, ServiceDiscovery


class AIInferenceAgent:
    def __init__(self, wallet: Keypair):
        self.config = X402Config(network="devnet")
        self.client = SolanaX402Client(wallet, self.config)
        self.discovery = ServiceDiscovery()

    async def find_inference_service(self, max_price: float = 0.05):
        services = await self.discovery.discover(
            query="inference",
            category="AI & ML",
            max_price=max_price,
            min_reputation=0.7,
            limit=1
        )

        if not services:
            raise Exception(f"No inference services found under {max_price} USD")

        return services[0]

    async def run_inference(self, prompt: str, model: str = "llama-3.1-8b"):
        print(f"Finding inference service (model: {model})...")
        service = await self.find_inference_service()

        print(f"Using service: {service.name}")
        print(f"Price: {service.pricing.get('amount')} {service.pricing.get('asset')}")

        print(f"\nRunning inference...")
        result = await self.client.post(
            service.resource,
            data={
                "messages": [
                    {"role": "user", "content": prompt}
                ],
                "model": model,
            }
        )

        return result

    async def close(self):
        await self.client.close()
        await self.discovery.close()


async def main():
    private_key_path = os.getenv("X402_WALLET_PATH", "wallet.json")

    try:
        import json
        with open(private_key_path, "r") as f:
            secret_key = json.load(f)
            wallet = Keypair.from_bytes(bytes(secret_key))
    except FileNotFoundError:
        print("Error: Wallet file not found")
        print("Set X402_WALLET_PATH environment variable or create wallet.json")
        return

    agent = AIInferenceAgent(wallet)

    try:
        balance = await agent.client.get_balance()
        print(f"Wallet: {agent.client.get_wallet_address()}")
        print(f"Balance: {balance} SOL\n")

        prompts = [
            "Explain quantum computing in simple terms",
            "What are the top 3 programming languages in 2024?",
            "How does blockchain work?",
        ]

        for i, prompt in enumerate(prompts, 1):
            print(f"\n{'='*60}")
            print(f"Query {i}: {prompt}")
            print('='*60)

            try:
                result = await agent.run_inference(prompt)

                if "choices" in result:
                    response = result["choices"][0]["message"]["content"]
                    print(f"\nResponse:\n{response}")

                    if "usage" in result:
                        tokens = result["usage"]["total_tokens"]
                        print(f"\nTokens used: {tokens}")

            except Exception as e:
                print(f"Inference failed: {e}")

            await asyncio.sleep(1)

        final_balance = await agent.client.get_balance()
        print(f"\n{'='*60}")
        print(f"Final balance: {final_balance} SOL")
        print(f"Total spent: {balance - final_balance} SOL")

    finally:
        await agent.close()


if __name__ == "__main__":
    asyncio.run(main())

File: packages/sdk/rust/examples/ai_inference.rs
use serde_json::json;
use solana_sdk::signature::Keypair;
use x402_upl::{SolanaX402Client, X402Config, ServiceDiscovery, DiscoveryConfig};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("x402-upl Rust SDK - AI Inference Example\n");

    let wallet = Keypair::new();

    let config = X402Config {
        network: "devnet".to_string(),
        rpc_url: Some("https://api.devnet.solana.com".to_string()),
        ..Default::default()
    };

    let client = SolanaX402Client::new(wallet, config)?;

    println!("Wallet: {}", client.get_wallet_address());

    let balance = client.get_balance("SOL")?;
    println!("Balance: {} SOL\n", balance);

    if balance < 0.01 {
        eprintln!("Error: Insufficient balance for testing");
        eprintln!("Fund your wallet:");
        eprintln!("  solana airdrop 1 {} --url devnet", client.get_wallet_address());
        std::process::exit(1);
    }

    println!("Finding AI inference service...");
    let discovery_config = DiscoveryConfig::default();
    let discovery = ServiceDiscovery::new(discovery_config)?;

    let services = discovery
        .discover(
            Some("inference"),
            Some("AI & ML"),
            Some(0.05),
            Some(0.7),
            None,
            1
        )
        .await?;

    if services.is_empty() {
        eprintln!("No inference services found");
        std::process::exit(1);
    }

    let service = &services[0];
    println!("Using service: {}", service.name);

    if let Some(pricing) = &service.pricing {
        println!("Price: {} {}\n", pricing.amount, pricing.asset);
    }

    let prompts = vec![
        "Explain quantum computing in simple terms",
        "What are the top 3 programming languages in 2024?",
        "How does blockchain work?",
    ];

    for (i, prompt) in prompts.iter().enumerate() {
        println!("{}. Query: {}", i + 1, prompt);
        println!("{}", "=".repeat(60));

        let request_body = json!({
            "messages": [
                {"role": "user", "content": prompt}
            ],
            "model": "llama-3.1-8b"
        });

        match client.post::<serde_json::Value>(&service.resource, Some(request_body), None).await {
            Ok(result) => {
                if let Some(choices) = result.get("choices") {
                    if let Some(first_choice) = choices.get(0) {
                        if let Some(message) = first_choice.get("message") {
                            if let Some(content) = message.get("content") {
                                println!("\nResponse:\n{}\n", content);
                            }
                        }
                    }
                }

                if let Some(usage) = result.get("usage") {
                    if let Some(tokens) = usage.get("total_tokens") {
                        println!("Tokens used: {}\n", tokens);
                    }
                }
            }
            Err(e) => {
                eprintln!("Inference failed: {}\n", e);
            }
        }

        tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
    }

    let final_balance = client.get_balance("SOL")?;
    println!("{}", "=".repeat(60));
    println!("Final balance: {} SOL", final_balance);
    println!("Total spent: {} SOL", balance - final_balance);

    Ok(())
}
